File: apps/csharp/tests/Ch04/ThreadBasicsTests.cs
using GrokkingConcurrency.CSharp.Ch04;
using System.Collections.Concurrent;

namespace GrokkingConcurrency.CSharp.Tests.Ch04;

/// <summary>
/// スレッド基本操作のテスト
/// </summary>
public class ThreadBasicsTests
{
    [Fact]
    public async Task ShouldCreateAndRunThread()
    {
        var results = new ConcurrentBag<string>();
        using var semaphore = new SemaphoreSlim(0, 1);

        var thread = ThreadBasics.CreateWorkerThread(() =>
        {
            results.Add("task executed");
            semaphore.Release();
        });

        thread.Start();
        await semaphore.WaitAsync(TimeSpan.FromSeconds(5));

        Assert.Single(results);
        Assert.Contains("task executed", results);
    }

    [Fact]
    public async Task ShouldRunMultipleThreadsConcurrently()
    {
        var results = new ConcurrentBag<int>();
        const int threadCount = 5;
        using var countdown = new CountdownEvent(threadCount);

        var threads = ThreadBasics.CreateMultipleWorkers(threadCount, i =>
        {
            results.Add(i);
            countdown.Signal();
        });

        foreach (var thread in threads)
        {
            thread.Start();
        }

        countdown.Wait(TimeSpan.FromSeconds(5));

        Assert.Equal(threadCount, results.Count);
        Assert.Equal(new[] { 0, 1, 2, 3, 4 }, results.OrderBy(x => x));
    }

    [Fact]
    public void ShouldSetThreadName()
    {
        var thread = ThreadBasics.CreateNamedThread("TestWorker", () => { });

        Assert.Equal("TestWorker", thread.Name);
    }

    [Fact]
    public void ShouldWaitForThreadCompletion()
    {
        var results = new ConcurrentBag<string>();

        var thread = ThreadBasics.CreateWorkerThread(() =>
        {
            Thread.Sleep(100);
            results.Add("completed");
        });

        thread.Start();
        Assert.Empty(results);

        thread.Join();
        Assert.Single(results);
        Assert.Contains("completed", results);
    }

    [Fact]
    public void ShouldGetActiveThreadCount()
    {
        var count = ThreadBasics.GetActiveThreadCount();

        Assert.True(count >= 1);
    }
}

File: apps/fsharp/tests/ThreadBasicsTests.fs
module ThreadBasicsTests

open Xunit
open System.Collections.Concurrent
open GrokkingConcurrency.Ch04.ThreadBasics

[<Fact>]
let ``createWorker creates a thread that runs the given task`` () =
    let mutable executed = false
    let thread = createWorker "TestWorker" (fun () -> executed <- true)
    thread.Start()
    thread.Join()
    Assert.True(executed)

[<Fact>]
let ``createWorker sets thread name`` () =
    let thread = createWorker "MyWorker" (fun () -> ())
    Assert.Equal("MyWorker", thread.Name)

[<Fact>]
let ``createMultipleWorkers creates specified number of threads`` () =
    let threads = createMultipleWorkers 5 (fun _ -> ())
    Assert.Equal(5, List.length threads)

[<Fact>]
let ``createMultipleWorkers threads execute with correct indices`` () =
    let results = ConcurrentQueue<int>()
    let threads = createMultipleWorkers 3 (fun i -> results.Enqueue(i))
    threads |> List.iter (fun t -> t.Start())
    threads |> List.iter (fun t -> t.Join())
    let resultSet = results |> Seq.toList |> Set.ofList
    Assert.True((Set.ofList [0; 1; 2]) = resultSet)

[<Fact>]
let ``startAndJoinAll runs all threads`` () =
    let results = ConcurrentQueue<string>()
    let threads = createMultipleWorkers 3 (fun i -> results.Enqueue($"Worker-{i}"))
    startAndJoinAll threads
    Assert.Equal(3, results.Count)
